src/main/java/net/minecraftforge/registries/DeferredRegister.java

```diff
@@ -99,24 +156,111 @@ public void register(IEventBus bus)
+                    e.getKey().getRegistryObject().updateReference(reg);
+                }
             }
         }
     }
+
+    /**
+     * If the IForgeRegistry is null, gets it from the supplier. This is sure to be called from a classic register event.
+     * It is even guaranteed to be called from the Register Block event, since it is the first one called.
+     * This also makes sure that the IForgeRegistry is correct when updating the RegistryObjects.
+     *
+     * This is only useful for custom registries, when the {@link IForgeRegistry} does not exist when static init happens
+     */
+    private IForgeRegistry<T> getFromSupplier(){
+        if(type == null)
+        {
+            type = typeSup.get();
+        }
+        return type;
+    }
+
+    /**
+     * A way of lazily evaluating registry objects but in a nicer way than using {@link LazyOptional<RegistryObject>}
+     *
+     * Also a supplier, to mimic {@link RegistryObject} as best as possible
+     */
+    public static class LazyRegistryObject<T extends IForgeRegistryEntry<? super T>> implements Supplier<T>
+    {
+        private final LazyOptional<RegistryObject<T>> lazyRet;
+        private RegistryObject<T> ret;
+        private LazyRegistryObject(NonNullSupplier<RegistryObject<T>> supplier)
+        {
+            if(supplier == null)
+            {
+                throw new IllegalStateException("Supplier should not be null");
+            }
+            lazyRet = LazyOptional.of(supplier);
+        }
+
+        /**
+         * The interest of the LazyOptional is mostly for the lazy part.
+         * This is only called in {@link #addEntries} when the IForgeRegistry is definitely assigned.
+         * The RegistryObject is created using {@link RegistryObject#of(ResourceLocation, IForgeRegistry)}
+         * It is only created once from the passed in supplier, to not evaluate it each time with {@link #get()}
+         */
+        private RegistryObject<T> getRegistryObject()
+        {
+            if(ret == null)
+            {
+                ret = lazyRet.orElseThrow(()->new IllegalStateException("This deserves a better error message."));
+            }
+            return ret;
+        }
+
+        /**
+         * Mostly for convenience, so that there is no need to do get().get(), but also to act as a Supplier just like RegistryObject
+         */
+        @Override
+        public T get()
+        {
+            return getRegistryObject().get();
+        }
+
+        /**
+         * Helper function to cast, since it is needed in {@link #lazyRegister} to work with
+         * objects not directly implementing the type of this registry
+         * But now that i think about it, pretty useless...
+         */
+        @SuppressWarnings("unchecked")
+        private  <X extends IForgeRegistryEntry<? super X>> LazyRegistryObject<X> cast()
+        {
+            return (LazyRegistryObject<X>) this;
+        }
```

Comment:
I feel like this would be cleaner doing it like how forge does it for the registry object and just inlining it `lazyRegister`.